### ISSUE TITLE
Failed to activate the hyperclick package

### ISSUE THREAD

**Issue by JustDevZero** [facebooknuclide/hyperclick#4]
Posted on https://github.com/facebook/nuclide/issues/852 like the description details.

**Comment by raoxudong** [comment]
Failed to execute 'registerElement' on 'Document': Registration failed for type 'hyperclick-suggestion-list'. A type with that name is already registered.

**Comment by qhaaakaa** [comment]
Failed to activate the hyperclick package   Cannot find module 'react-for-atom'

**Comment by omih** [comment]
[Enter steps to reproduce:]

1. ...
2. ...

**Atom**: 1.15.0 x64
**Electron**: 1.3.13
**OS**: Mac OS X 10.12.3
**Thrown From**: [hyperclick](https://github.com/facebooknuclide/hyperclick) package 0.0.40


### Stack Trace

Failed to activate the hyperclick package

```
At Failed to execute 'registerElement' on 'Document': Registration failed for type 'hyperclick-suggestion-list'. A type with that name is already registered.

Error: Failed to execute 'registerElement' on 'Document': Registration failed for type 'hyperclick-suggestion-list'. A type with that name is already registered.
    at Error (native)
    at /packages/hyperclick/lib/SuggestionListElement.js:261:31)
    at Module._compile (/app.asar/src/native-compile-cache.js:109:30)
    at /app.asar/src/compile-cache.js:216:21)
    at Module.load (module.js:473:32)
    at tryModuleLoad (module.js:432:12)
    at Function.Module._load (module.js:424:3)
    at Module.require (module.js:483:17)
    at require (/app.asar/src/native-compile-cache.js:50:27)
    at /packages/hyperclick/lib/main.js:31:30)
    at Module._compile (/app.asar/src/native-compile-cache.js:109:30)
    at /app.asar/src/compile-cache.js:216:21)
    at Module.load (module.js:473:32)
    at tryModuleLoad (module.js:432:12)
    at Function.Module._load (module.js:424:3)
    at Module.require (module.js:483:17)
    at require (/app.asar/src/native-compile-cache.js:50:27)
    at Package.module.exports.Package.requireMainModule (/app.asar/src/package.js:796:27)
    at Package.module.exports.Package.activateNow (/app.asar/src/package.js:207:16)
    at /app.asar/src/package.js:190:32
    at Package.module.exports.Package.measure (/app.asar/src/package.js:96:15)
    at /app.asar/src/package.js:183:26
    at Package.module.exports.Package.activate (/app.asar/src/package.js:180:34)
    at PackageManager.module.exports.PackageManager.activatePackage (/app.asar/src/package-manager.js:550:34)
    at /app.asar/node_modules/settings-view/lib/package-manager.js:484:29
    at exit (/app.asar/node_modules/settings-view/lib/package-manager.js:99:16)
    at triggerExitCallback (/app.asar/src/buffered-process.js:322:11)
    at /app.asar/src/buffered-process.js:343:11
    at /app.asar/src/buffered-process.js:204:9)
    at emitOne (events.js:101:20)
    at Socket.emit (events.js:188:7)
    at Pipe._handle.close [as _onclose] (net.js:493:12)
```

### Commands

```
  8x -8:56.9.0 core:backspace (input.hidden-input)
     -8:28.4.0 application:new-file (ul.list-inline.tab-bar.inset-panel)
  2x -5:36.7.0 nuclide-home:toggle (atom-workspace.workspace.scrollbars-visible-always.theme-one-light-syntax.theme-one-light-ui)
```

### Non-Core Packages

```
atom-material-syntax 1.0.2 
chester-atom-syntax 0.1.1 
gl-light-syntax 2.2.0 
hyperclick 0.0.40 
nuclide 0.217.0 
nuclide-format-js 0.0.36 
script 3.14.1 
Termrk 1.0.0 
```

**Comment by atomi** [comment]
[Enter steps to reproduce:]

1. install go-plus package
2. ...

**Atom**: 1.19.0-beta0 x64
**Electron**: 1.6.9
**OS**: Linux Mint
**Thrown From**: [hyperclick](https://github.com/facebooknuclide/hyperclick) package 0.0.40


### Stack Trace

Failed to activate the hyperclick package

```
At Cannot read property 'getDomNode' of undefined

TypeError: Cannot read property 'getDomNode' of undefined
    at getLinesDomNode (/packages/hyperclick/lib/HyperclickForTextEditor.js:99:40)
    at addMouseListeners (/packages/hyperclick/lib/HyperclickForTextEditor.js:109:9)
    at HyperclickForTextEditor._setupMouseListeners (/packages/hyperclick/lib/HyperclickForTextEditor.js:115:7)
    at /packages/hyperclick/lib/HyperclickForTextEditor.js:60:10)
    at Hyperclick.observeTextEditor (/packages/hyperclick/lib/Hyperclick.js:85:37)
    at /packages/hyperclick/lib/Hyperclick.js:74:20
    at Workspace.observeTextEditors (/usr/share/atom-beta/resources/app/src/workspace.js:717:15)
    at /packages/hyperclick/lib/Hyperclick.js:62:51)
    at Object.activate (/packages/hyperclick/lib/main.js:55:16)
    at Package.module.exports.Package.activateNow (/usr/share/atom-beta/resources/app/src/package.js:253:25)
    at /usr/share/atom-beta/resources/app/src/package.js:225:38
    at Package.module.exports.Package.measure (/usr/share/atom-beta/resources/app/src/package.js:99:21)
    at /usr/share/atom-beta/resources/app/src/package.js:218:32
    at Package.module.exports.Package.activate (/usr/share/atom-beta/resources/app/src/package.js:215:40)
    at PackageManager.module.exports.PackageManager.activatePackage (/usr/share/atom-beta/resources/app/src/package-manager.js:645:40)
    at /usr/share/atom-beta/resources/app/node_modules/settings-view/lib/package-manager.js:535:35
    at exit (/usr/share/atom-beta/resources/app/node_modules/settings-view/lib/package-manager.js:128:22)
    at triggerExitCallback (/usr/share/atom-beta/resources/app/src/buffered-process.js:338:17)
    at /usr/share/atom-beta/resources/app/src/buffered-process.js:359:17
    at Socket.<anonymous> (/usr/share/atom-beta/resources/app/src/buffered-process.js:220:15)
    at emitOne (events.js:101:20)
    at Socket.emit (events.js:188:7)
    at Pipe._handle.close [as _onclose] (net.js:501:12)
```

### Commands

```
```

### Non-Core Packages

```
fold-search-results 1.0.5 
git-plus 7.9.3 
go-debug 1.4.3 
go-plus 5.5.5 
go-signature-statusbar 1.2.0 
hyperclick 0.0.40 
language-docker 1.1.8 
language-generic-config 1.3.0 
language-quicktemplate 1.0.0 
linter 2.1.4 
linter-ui-default 1.6.1 
```

**Comment by OlegSmelov** [comment]
Since the standalone `hyperclick` package [is probably not getting updated](https://github.com/facebook/nuclide/issues/852#issuecomment-257353938), here's a workaround for `At Cannot read property 'getDomNode' of undefined`.

Apply [this patch](https://gist.github.com/OlegSmelov/0c5c5a281a55cec3d3d406c01cb4605e) to your locally installed package. The patch is a backport of https://github.com/facebook/nuclide/commit/9c0b53feb1682569c5a53c1b67a49eaf987d3fbb.

OS X / Linux instructions:

```bash
$ cd ~/.atom/packages/hyperclick
$ wget https://gist.githubusercontent.com/OlegSmelov/0c5c5a281a55cec3d3d406c01cb4605e/raw -O get-dom-node.patch
$ patch -p 1 < get-dom-node.patch
$ rm get-dom-node.patch # optional cleanup
```

Then restart Atom.

**Comment by SkunkShow** [comment]
@OlegSmelov  how do I fix this in Windows? What am i suposed to do?

**Comment by OlegSmelov** [comment]
@NunoMouraz I haven't used Windows for development in a while, but I think you could apply the patch manually. To edit your locally installed package, you can use this button:

![hyperclick-edit](https://user-images.githubusercontent.com/544537/27271997-a2e44494-54d0-11e7-8200-97164c324989.png)

Then go through the changes [in the patch](https://gist.github.com/OlegSmelov/0c5c5a281a55cec3d3d406c01cb4605e) and manually change the lines of code as shown. Not very convenient, I know.

Alternatively, if you have BashOnWindows, Cygwin or MSYS2 installed, you could run the commands for OS and Linux, except for the first one, since your package will be installed somewhere else.

**Comment by apaleslimghost** [comment]
So I had been planning to release and maintain a forked standalone version, but the terms of the non-free [Nuclide licence](https://github.com/facebooknuclide/hyperclick/blob/master/LICENSE) prevent me from doing that.

@zertosh, @nmote, @matthewwithanm, @hansonw, @mostafaeweda, @peterhal, @jgebhardt, if you're not going to maintain standalone Hyperclick, can you _please please please_ relicense it under a permissive licence so the community can. 180,000 people depend on this package. Not everybody wants to install all of Nuclide.

**Comment by hansonw** [comment]
Good news - we've started to publish the language-services portions of Nuclide under a separate package (https://atom.io/packages/atom-ide-ui) which contains Hyperclick. (You may want to give that a try, though it's still not ready for public release). `atom-ide-ui` is licensed under BSD with a patents clause, which is the same license that React uses.

That means we can also update this copy to have the better licensing - I'll try to do that sometime this week :)

**Comment by hansonw** [comment]
I just published v0.1.0 which has the fix. Disclaimer: this repo is now just a bundled version of the `atom-ide-ui` code, but that means that we can keep maintaining this repo for free :D
